### DISCUSSION TITLE
How to get a custom metafield from Shopify into Statamic?

### DISCUSSION THREAD

**Discussion by jensolafkoch** [statamic-rad-pack/shopify#132] [Help]
I already read the two other discussions re metafields but – as a newbie to both Shopify and the Shopify addon – I wonder if I can import my metafield data into Statamic and if so, how. 

I couldn't find anything in the docs re metafields as well.

**Comment by jackabox** [comment]
Hey @jensolafkoch, it's not something I've coded in to easily supply a method to sync it, but you could extend the importer to grab the metafields on import and map them to the appropriate fields.

**Reply by jensolafkoch** [reply]
Thanks, @jackabox, I'll have a look into that!

I wonder: Do I have to extend the importer as well to get the field product number (SKU) – which is different from the internal Shopify Product ID. 

You only list the Product ID here: [https://statamic-shopify.jackwhiting.co.uk/cms/content/#product-content](https://statamic-shopify.jackwhiting.co.uk/cms/content/#product-content)

**Reply by jensolafkoch** [reply]
@jackabox I tried to locate the importer but couldn't find a class in your add-on that I could extend to also query metafields. If I look with ray() at the raw product data I get from Shopify, like here ...

```
trait FetchAllProducts
{
    public function fetchProducts()
    {
        $shopify = new ShopifySDK();

        $productResource = $shopify->Product();
        $products = $productResource->get(['limit' => config('shopify.api_limit')]);
        ray($products);
```

... there are no metafields to be found. I still wonder how I can get those from within PHP to save the data to the product entry.

Could you please elaborate a bit where exactly I can extend your import process? 

Or do you meant extending `PHPShopify\ShopifySDK` or anything in `phpclassic/php-shopify`?

Many thanks!!

**Reply by jackabox** [reply]
The product ID is in the local files, metafields are not as easy to get as just grabbing them on the imported product. For Metafields, you'd need to import them specifically using some variation fo the following https://shopify.dev/api/examples/metafields#step-2-retrieve-metafields

**Reply by jensolafkoch** [reply]
Alright, thanks, will look into this.

**Reply by jensolafkoch** [reply]
@jackabox 

I dug a bit deeper, and it looks like there is a REST API one can use to access Shopify resources, including the Metafield resource ([https://shopify.dev/api/admin-rest/2022-07/resources/metafield](https://shopify.dev/api/admin-rest/2022-07/resources/metafield)).

Using a package like Laravel Shopify it should be possible to access such endpoints ([Laravel Shopify: Supported Methods](https://www.notion.so/Laravel-Shopify-424709fa6dd84defb0879ab8a1cb29fa#7667e0e66ab24dab81d233009bb2433f)), what do you think from your Shopify experience?

*

Problem: Both your Addon as well as this package Laravel Shopify use shopify.php as their config file name. I'm not sure how to deal with such a conflict.

**Reply by jackabox** [reply]
You don't need another package to do it, you can extend the current implementation and use the REST API endpoint. Here: https://shopify.dev/api/admin-rest/2022-07/resources/metafield#get-metafields?metafield[owner-id]=674387490&metafield[owner-resource]=article.

You'd need to write a new function which would run and grab the metafields based on the `owner-resource` being a product, and the `owner-id` being the product id. 

It's not something I've got supported in the core of this package as it basically defeats the point of using the products within Statamic. I'm on the side that the "metafields"/"custom fields" should just be set in statamic by adding them to the blueprint. Keeping the core product information only sync'd between the two.

Not sure what your end goal is here, but the far easier way is letting the product owner add the fields within Statamic.

**Comment by jensolafkoch** [comment]
Thanks again, I'll check this out!

Our client wants an internal custom article number and a custom release date (it's a publisher), different from the Shopify published_at date, both kept central in Shopify, not in Statamic, to have a single source of truth. That's the rationale here, otherwise we would be happy to add custom fields in Statamic. But maybe we will or have to re-consider ...

**Reply by jackabox** [reply]
Hey @jensolafkoch, in that case you'd need to extend the importer with metafields. This is doable, but just not included in the core functionality of the add-on at the moment. 

If you cannot get it working from your side, then do reach out, and we may be able to discuss something :)

**Reply by jensolafkoch** [reply]
Hi @jackabox,

for the time being I added an Action for product entries to get the metafields via Rest API and mapped and saved them to blueprint fields. 
So at least I know how to fetch them.

I'm not sure yet how to automate that, but I hope to hook somewhere into the process after product(s) have been imported by the addon. Is this what you meant? I'm still quite new to Rest APIs and queues and webhooks, at least in a practical sense, but in programming everything is or was new at some point ...

Thanks for you offer, if I really get stuck to work something out...!! I still hope to get through the jungle myself, it's still the best way to learn things, and we have another project with Shopify and your add-on coming up (if i get it to work somehow :-) ), so I'll try to dig deeper and understand the details ...

**Reply by jensolafkoch** [reply]
I've managed to add a callback in the `boot()` method of `AppServiceProvider` to hook after every finished job, detect the `ImportSingleProductJob` and then get the metafields for the imported product immediately thereafter.

```
public function boot()
{
        Queue::after(function (JobProcessed $event) {

        // $event->connectionName
        // $event->job
        // $event->exception

        $payload = $event->job->payload();
        $soughtJob = "Jackabox\Shopify\Jobs\ImportSingleProductJob"; // Statamic-Addon: Shopify
        $jobDisplayName = Arr::get($payload, 'displayName');

        if(Str::endsWith($jobDisplayName, $soughtJob)) {

            $job = unserialize($payload['data']['command']);
            $productId = Arr::get($job->data, 'id');

            MetafieldsController::getMetafields($productId);

        }

    });
}
```

As far as I understand your sources you don't fire any own event after the job has been processed? That would be a nice touch as the callback body would be a one-liner then:

```
Queue::after(function (ImportSingleProductJobProcessed $event) {

          MetafieldsController::getMetafields($productId);

      }

  });
```

**Comment by ryanmitchell** [comment]
This will be included in the next release.
